File: backend/routers/alerts.py
"""Cross-account Resource Monitor aggregation for the Alerts page."""
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, Query

from backend.lib import cost_queries, db
from backend.security import get_current_user

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


def _disabled_client(connection: dict, detail: str) -> dict:
    return {
        "id": connection["id"],
        "name": connection["name"],
        "account_identifier": connection["account_identifier"],
        "status": "disabled",
        "error": detail,
        "note": None,
        "monitors": [],
        "max_quota_used_pct": None,
    }


def _fetch_client(connection: dict, credentials: dict) -> dict:
    payload = cost_queries.resource_monitors(credentials)
    monitors = payload.get("items") or []
    note = payload.get("note")

    # _run_show intentionally degrades Snowflake failures to a note. Alerts needs
    # to surface those accounts as not monitored instead of silently calling them healthy.
    if note and not monitors and str(note).startswith("Sem acesso"):
        return _disabled_client(connection, str(note))

    percentages = [
        float(item["quota_used_pct"])
        for item in monitors
        if item.get("quota_used_pct") is not None
    ]
    monitors.sort(
        key=lambda item: (
            item.get("quota_used_pct") is not None,
            float(item.get("quota_used_pct") or 0),
        ),
        reverse=True,
    )
    return {
        "id": connection["id"],
        "name": connection["name"],
        "account_identifier": connection["account_identifier"],
        "status": "active",
        "error": None,
        "note": note,
        "monitors": monitors,
        "max_quota_used_pct": max(percentages) if percentages else 0,
    }


@router.get("")
def alerts_resource_monitors(
    connection_id: list[int] | None = Query(None),
    user: dict = Depends(get_current_user),
):
    """Fetch accessible accounts independently so one failure never blocks the rest."""
    accessible = db.list_connections_for_user(user)
    if connection_id:
        requested = set(connection_id)
        accessible_ids = {row["id"] for row in accessible}
        if not requested.issubset(accessible_ids):
            raise HTTPException(status_code=403, detail="Sem permissão para uma das conexões.")
        connections = [row for row in accessible if row["id"] in requested]
    else:
        connections = accessible

    clients_by_id: dict[int, dict] = {}
    jobs: list[tuple[dict, dict]] = []
    for connection in connections:
        try:
            jobs.append((connection, db.get_connection_credentials(connection["id"])))
        except Exception as exc:  # noqa: BLE001
            clients_by_id[connection["id"]] = _disabled_client(connection, str(exc))

    if jobs:
        with ThreadPoolExecutor(max_workers=min(8, len(jobs))) as executor:
            futures = {
                executor.submit(_fetch_client, connection, credentials): connection
                for connection, credentials in jobs
            }
            for future in as_completed(futures):
                connection = futures[future]
                try:
                    clients_by_id[connection["id"]] = future.result()
                except Exception as exc:  # noqa: BLE001
                    clients_by_id[connection["id"]] = _disabled_client(connection, str(exc))

    clients = [clients_by_id[row["id"]] for row in connections]
    clients.sort(
        key=lambda client: (
            client["status"] == "active",
            float(client.get("max_quota_used_pct") or 0),
        ),
        reverse=True,
    )
    active = sum(client["status"] == "active" for client in clients)
    disabled = len(clients) - active
    critical = sum(
        client["status"] == "active"
        and float(client.get("max_quota_used_pct") or 0) >= 70
        for client in clients
    )
    return {
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "total_connections": len(clients),
        "active_connections": active,
        "disabled_accounts": disabled,
        "critical_clients": critical,
        "clients": clients,
    }

File: frontend/src/api/alerts.ts
import { api } from "./client";

export type AlertMonitor = {
  name: string;
  quota_used_pct: number | null;
  credit_quota: number | null;
  used_credits: number | null;
  remaining_credits: number | null;
  level: string;
  frequency: string;
  warehouses: string[];
  start_time: string | null;
};

export type AlertClient = {
  id: number;
  name: string;
  account_identifier: string;
  status: "active" | "disabled";
  error: string | null;
  note: string | null;
  monitors: AlertMonitor[];
  max_quota_used_pct: number | null;
};

export type AlertsResponse = {
  fetched_at: string;
  total_connections: number;
  active_connections: number;
  disabled_accounts: number;
  critical_clients: number;
  clients: AlertClient[];
};

export const ALERTS_BADGE_EVENT = "snow-portal:alerts-badge";

let cached: AlertsResponse | null = null;
let pending: Promise<AlertsResponse> | null = null;

export function announceAlertsBadge(count: number) {
  window.dispatchEvent(new CustomEvent(ALERTS_BADGE_EVENT, { detail: count }));
}

export async function getAlerts(
  options: { force?: boolean; connectionIds?: number[] } = {},
): Promise<AlertsResponse> {
  const ids = options.connectionIds;
  const isFullRequest = !ids?.length;

  if (isFullRequest && !options.force && cached) return cached;
  if (isFullRequest && !options.force && pending) return pending;

  const query = ids?.length
    ? `?${ids.map((id) => `connection_id=${encodeURIComponent(id)}`).join("&")}`
    : "";
  const request = api<AlertsResponse>(`/api/alerts${query}`);
  if (isFullRequest) pending = request;

  try {
    const response = await request;
    if (isFullRequest) {
      cached = response;
      announceAlertsBadge(response.critical_clients);
    }
    return response;
  } finally {
    if (isFullRequest && pending === request) pending = null;
  }
}
